lib/ddtrace/tracer.rb

```diff
@@ -448,11 +450,16 @@ def deactivate_priority_sampling!(base_sampler = nil)
       @sampler = base_sampler || Datadog::AllSampler.new if @sampler.is_a?(PrioritySampler)
     end
 
+    def register_shutdown!
+      at_exit { shutdown! }
```

Comment:
should we do this in the worker?

when the worker thread exits, it stops itself?

Reply:
I thought about this, but decided against it, because the Tracer might have more shutdown tasks than just the worker (e.g. if we have multiple threads, one for traces, one for run time metrics, etc...) and thought it better to make this hook broadly useful by promoting it.

What I don't really like is that for each Tracer object, we're registering an `at_exit` hook... this might not be a good idea, because we could be creating a lot of callbacks, and once we create them, it doesn't appear we can unregister them. Additionally, these callbacks could cause any referenced objects to remain uncollectable by the GC.

Originally I thought it would be best to automatically ensure any created Tracer would be shutdown automatically. But now I'm considering maybe we should do this only once at a more global level e.g. `at_exit { Datadog.tracer.shutdown! }` out-of-the-box, and if you're creating your own tracer instances, it's then your responsibility to define any `at_exit` hooks. Doing so should cover almost all users (as most do not use any tracer instance other than the default) but not be unreasonably burdensome for those who do more specialized implementations.

Reply:
I was on the fence on this one, but after realizing that
```ruby
10000.times do { Datadog::Tracer.new }
```
will create memory that will never be freed up, I'm in favour of `at_exit { Datadog.tracer.shutdown! }` only out of the box.

Reply:
We could also
```ruby
register_shutdown! if options.fetch(:register_shutdown)
```
in the initializer, defaulting to false.

Reply:
I'd rather not have this disabled by default, we should always be sure our workers have a chance to shutdown properly

Reply:
Sorry, what I meant to say is that if any user calls `Datadog::Tracer.new`, it should not register the hook by default.

We would always call it with `Datadog::Tracer.new(register_shutdown: true)` ourselves, which currently is only once per application.

Reply:
Interesting.

Reply:
I'm indifferent to whether we make it configurable, but slightly lean against, only because it's one more feature we have to support in our public API. I think if you're creating your own tracers, it's not anymore burdensome to add `at_exit { tracer.shutdown! }` than it is `Tracer.new(register_shutdown: true)`.

That said I'll do whichever if anyone has strong feelings about either one.

Reply:
Smaller API footprint sounds like a reasonable argument to me.